File: scripts/merge_empty_arabic.py
#!/usr/bin/env python3
# Example fix: merges an entry with empty "arabic" into the previous segment's
# translation/transcription, then deletes the empty entry and reindexes.
import argparse
import json
from pathlib import Path


def merge_into_previous(prev, current, field):
    cur_val = current.get(field, "")
    if cur_val is None:
        cur_val = ""
    cur_val = str(cur_val)
    if cur_val.strip() == "":
        return

    prev_val = prev.get(field, "")
    if prev_val is None:
        prev_val = ""
    prev_val = str(prev_val)

    if prev_val.strip() == "":
        prev[field] = cur_val.strip()
    else:
        prev[field] = prev_val.rstrip() + " " + cur_val.lstrip()


def merge_file(path: Path, dry_run: bool) -> tuple[int, int, bool]:
    data = json.loads(path.read_text(encoding="utf-8"))
    data.sort(key=lambda item: item.get("index", 0))

    new_data = []
    removed = 0
    no_prev = 0

    for item in data:
        arabic = item.get("arabic", "")
        if arabic is None:
            arabic = ""
        if str(arabic).strip() == "":
            removed += 1
            if new_data:
                prev = new_data[-1]
                merge_into_previous(prev, item, "translation")
                merge_into_previous(prev, item, "transcription")
            else:
                no_prev += 1
            continue
        new_data.append(item)

    if removed:
        for idx, item in enumerate(new_data, start=1):
            item["index"] = idx
        if not dry_run:
            path.write_text(
                json.dumps(new_data, ensure_ascii=False, indent=2),
                encoding="utf-8",
            )
    return removed, no_prev, removed > 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description=(
            "Merge empty-arabic entries into the previous segment's "
            "translation/transcription, then delete and reindex."
        )
    )
    parser.add_argument(
        "--base-dir",
        default="public/surat/segmented/de/27",
        help="Directory containing segmented surah JSON files.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Compute changes without writing files.",
    )
    args = parser.parse_args()

    script_root = Path(__file__).resolve().parent.parent
    base_dir = Path(args.base_dir)
    if not base_dir.is_absolute():
        base_dir = script_root / base_dir

    files = sorted(base_dir.glob("*.json"))
    total_removed = 0
    total_no_prev = 0
    files_changed = 0

    for path in files:
        removed, no_prev, changed = merge_file(path, args.dry_run)
        total_removed += removed
        total_no_prev += no_prev
        if changed:
            files_changed += 1

    print(f"files_changed={files_changed}")
    print(f"removed_entries={total_removed}")
    print(f"no_previous_entries={total_no_prev}")
    if args.dry_run and files_changed:
        print("dry-run: no files were written")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/merge_empty_arabic.mjs
#!/usr/bin/env node

import { readdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const DEFAULT_BASE_DIR = "public/surat/segmented/de/27";

function printHelp() {
  console.log(
    [
      "Merge empty-arabic entries into the previous segment's translation/transcription, then delete and reindex.",
      "",
      "Usage:",
      "  node scripts/merge_empty_arabic.mjs [--base-dir <path>] [--dry-run]",
      "",
      "Options:",
      `  --base-dir <path>  Directory containing segmented surah JSON files. (default: ${DEFAULT_BASE_DIR})`,
      "  --dry-run          Compute changes without writing files.",
      "  -h, --help         Show this help message.",
    ].join("\n"),
  );
}

function parseArgs(argv) {
  let baseDir = DEFAULT_BASE_DIR;
  let dryRun = false;

  for (let i = 2; i < argv.length; i += 1) {
    const arg = argv[i];

    if (arg === "--dry-run") {
      dryRun = true;
      continue;
    }

    if (arg === "--base-dir") {
      const value = argv[i + 1];
      if (!value) {
        throw new Error("Missing value for --base-dir");
      }
      baseDir = value;
      i += 1;
      continue;
    }

    if (arg.startsWith("--base-dir=")) {
      const value = arg.slice("--base-dir=".length);
      if (!value) {
        throw new Error("Missing value for --base-dir");
      }
      baseDir = value;
      continue;
    }

    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    }

    throw new Error(`Unknown argument: ${arg}`);
  }

  return { baseDir, dryRun };
}

function mergeIntoPrevious(prev, current, field) {
  let curVal = current[field] ?? "";
  curVal = String(curVal);
  if (curVal.trim() === "") {
    return;
  }

  let prevVal = prev[field] ?? "";
  prevVal = String(prevVal);

  if (prevVal.trim() === "") {
    prev[field] = curVal.trim();
  } else {
    prev[field] = `${prevVal.trimEnd()} ${curVal.trimStart()}`;
  }
}

async function mergeFile(filePath, dryRun) {
  const raw = await readFile(filePath, "utf8");
  const data = JSON.parse(raw);
  data.sort((a, b) => (a?.index ?? 0) - (b?.index ?? 0));

  const newData = [];
  let removed = 0;
  let noPrev = 0;

  for (const item of data) {
    const arabic = String(item?.arabic ?? "");
    if (arabic.trim() === "") {
      removed += 1;
      if (newData.length > 0) {
        const prev = newData[newData.length - 1];
        mergeIntoPrevious(prev, item, "translation");
        mergeIntoPrevious(prev, item, "transcription");
      } else {
        noPrev += 1;
      }
      continue;
    }

    newData.push(item);
  }

  if (removed > 0) {
    for (let idx = 0; idx < newData.length; idx += 1) {
      newData[idx].index = idx + 1;
    }
    if (!dryRun) {
      await writeFile(filePath, JSON.stringify(newData, null, 2), "utf8");
    }
  }

  return { removed, noPrev, changed: removed > 0 };
}

async function main() {
  const { baseDir, dryRun } = parseArgs(process.argv);

  const scriptPath = fileURLToPath(import.meta.url);
  const scriptRoot = path.resolve(path.dirname(scriptPath), "..");
  const resolvedBaseDir = path.isAbsolute(baseDir)
    ? baseDir
    : path.join(scriptRoot, baseDir);

  const entries = await readdir(resolvedBaseDir, { withFileTypes: true });
  const files = entries
    .filter((entry) => entry.isFile() && entry.name.endsWith(".json"))
    .map((entry) => path.join(resolvedBaseDir, entry.name))
    .sort((a, b) => a.localeCompare(b));

  let totalRemoved = 0;
  let totalNoPrev = 0;
  let filesChanged = 0;

  for (const filePath of files) {
    const { removed, noPrev, changed } = await mergeFile(filePath, dryRun);
    totalRemoved += removed;
    totalNoPrev += noPrev;
    if (changed) {
      filesChanged += 1;
    }
  }

  console.log(`files_changed=${filesChanged}`);
  console.log(`removed_entries=${totalRemoved}`);
  console.log(`no_previous_entries=${totalNoPrev}`);
  if (dryRun && filesChanged) {
    console.log("dry-run: no files were written");
  }
}

main().catch((error) => {
  console.error(error?.message ?? String(error));
  process.exit(1);
});
